// SPDX-FileCopyrightText: 2025 Carnegie Mellon University - Living Edge Lab
//
// SPDX-License-Identifier: GPL-2.0-only

syntax = "proto3";
package steeleagle.protocol.messages.result;

import "common.proto";
import "google/protobuf/timestamp.proto";

/*
 * Bounding box associated with an object detection.
 * 
 * Defines the upper left and lower right corners of a detected object
 * in an image frame. Origin (0,0) is the top left corner of the input image.
 * (image_height, image_width) is the bottom right corner.
 * Also the class and confidence threshold associated with the box.
 */
message BoundingBox {
  double y_min = 1; // minimum y offset percentage with respect to image size [0.0-1.0]
  double x_min = 2; // minimum x offset percentage with respect to image size [0.0-1.0]
  double y_max = 3; // maximum y offset percentage with respect to image size [0.0-1.0]
  double x_max = 4; // maximum x offset percentage with respect to image size [0.0-1.0]
}

/*
 * HSV values generated by a color filter.
 * 
 * Color filter represented by hue, saturation, and value Uses OpenCV ranges defined [here](https://opencv.org/blog/color-spaces-in-opencv/).
 */
message HSV {
  uint32 h = 1; // hue range is [0,179]
  uint32 s = 2; // saturation range is [0,255]
  uint32 v = 3; // value range is [0,255]
}

// Object detection generated by a model.
message Detection {
  uint64 detection_id = 1; // can be multiple objects per frame
  string class_name = 2; // class name of detection
  double score = 3; // confidence score
  BoundingBox bbox = 4; // bounding box
}

// List of object detections.
message DetectionResult {
  repeated Detection detections = 1; // list of detections
  uint64 frame_id = 2; // frame corresponding to these detections
}

// Avoidance result generated by an avoidance model.
message AvoidanceResult {
  double actuation_vector = 1; // actuation vector towards safe area
}

// SLAM result generated by a SLAM model.
message SLAMResult {
  oneof position { // slam estimated position
    steeleagle.protocol.common.Position relative_position = 1; // relative position estimate relative to SLAM initialization
    steeleagle.protocol.common.Location global_position = 2; // global position estimate
  }
}

// Compute result generated by a compute server.
message ComputeResult {
  google.protobuf.Timestamp timestamp = 1; // inference timestamp
  string engine_name = 2; // engine that generated the result
  oneof type { // type of detection
    DetectionResult detection_result = 3; // object detection
    AvoidanceResult avoidance_result = 4; // avoidance directive
    SLAMResult slam_result = 5; // SLAM position estimate
    string generic_result = 6; // JSON result
  }
}

/*
 * Compute results generated by datasink modules
 */
message FrameResult {
  string type = 1; // result type(s)
  uint64 frame_id = 2; // for correlation
  repeated ComputeResult result = 3; // list of generated results
}
